crates/storage/src/connection/event.rs

```diff
@@ -145,455 +105,159 @@ pub(super) fn get_events<K: KeyFilter>(
-        if events_by_key_filter < KEY_FILTER_UPPER_BOUND {
-            tracing::trace!(
-                %events_by_key_filter,
-                "Partial queries for number of events done"
-            );
-            return Ok(QueryStrategy::KeysFirst);
-        }
-    } else {
-        // we have no key filter at all
-        return Ok(QueryStrategy::BlockRangeFirst);
-    }
+                if event.keys.len() < keys.len() {
+                    return false;
+                }
 
-    let events_in_block_range =
-        number_of_events_in_block_range(tx, from_block, to_block, contract_address)?;
-
-    tracing::trace!(
-        ?events_in_block_range,
-        ?events_by_key_filter,
-        "Partial queries for number of events done"
-    );
-
-    let weighted_events_by_key_filter =
-        events_by_key_filter.map(|n| n.saturating_mul(KEY_FILTER_WEIGHT));
-
-    if events_in_block_range
-        .or(weighted_events_by_key_filter)
-        .is_some()
-    {
-        let cost = std::cmp::min(
-            events_in_block_range.unwrap_or(usize::MAX),
-            weighted_events_by_key_filter.unwrap_or(usize::MAX),
-        );
+                event
+                    .keys
+                    .iter()
+                    .zip(keys.iter())
+                    .all(|(key, filter)| filter.is_empty() || filter.contains(key))
+            })
+            .skip_while(|_| {
+                let skip = offset > 0;
+                offset = offset.saturating_sub(1);
+                skip
+            })
+            .take(events_required)
+            .map(|(event, tx_hash)| EmittedEvent {
+                data: event.data.clone(),
+                keys: event.keys.clone(),
+                from_address: event.from_address,
+                block_hash: block_header.hash,
+                block_number: block_header.number,
+                transaction_hash: tx_hash,
+            });
 
-        if cost > KEY_FILTER_COST_LIMIT {
-            return Err(EventFilterError::TooManyMatches.into());
-        }
+        emitted_events.extend(events);
     }
 
-    let strategy = match (events_in_block_range, events_by_key_filter) {
-        (None, None) => QueryStrategy::BlockRangeFirst,
-        (None, Some(_)) => QueryStrategy::KeysFirst,
-        (Some(_), None) => QueryStrategy::BlockRangeFirst,
-        (Some(events_in_block_range), Some(_)) => {
-            if events_in_block_range
-                > weighted_events_by_key_filter
-                    .expect("Unwrap is safe because events_by_key_filter is some")
-            {
-                QueryStrategy::KeysFirst
-            } else {
-                QueryStrategy::BlockRangeFirst
-            }
-        }
-    };
+    let is_last_page = emitted_events.len() <= filter.page_size;
+    emitted_events.truncate(filter.page_size);
 
-    Ok(strategy)
+    Ok(PageOfEvents {
+        events: emitted_events,
+        is_last_page,
+    })
 }
 
-fn number_of_events_in_block_range(
+fn load_bloom(
```

Comment:
Could be made `BloomFilter::load`

Reply:
Up to you - I don't really mind. In a way the cache is tied to bloom so it could make sense to move. But really just a style suggestion so not overly fussed.